File: backend/internal/service/platform_catalog.go
package service

import (
	"sort"

	"github.com/Wei-Shaw/sub2api/internal/domain"
)

// PlatformCatalog 是前端使用的平台清单：domain 平台清单加上多协议供应商的
// profile（接入模式、各协议默认基址、内置分流规则）。它被写成前端内置清单
// frontend/src/constants/platformCatalog.builtin.json（见
// TestFrontendBuiltinPlatformCatalogInSync），前端据此生成平台选项、展示名与
// 多协议账号表单，新登记的平台只需重新生成该文件，无需修改前端代码。
type PlatformCatalog struct {
	Platforms []PlatformCatalogEntry `json:"platforms"`
	// CompositePrecedence 是组合分组逐平台回退的顺序（先命中者生效）。
	CompositePrecedence []string `json:"composite_precedence"`
}

// PlatformCatalogEntry 描述一个具体平台；MultiProtocol 仅多协议 API Key 供应商有值。
type PlatformCatalogEntry struct {
	domain.PlatformSpec
	MultiProtocol *ProviderProfileInfo `json:"multi_protocol,omitempty"`
}

// ProviderProfileInfo 是 ProviderProfile 的对外形状。
type ProviderProfileInfo struct {
	DefaultMode   string          `json:"default_mode"`
	Routing       ProviderRouting `json:"routing"`
	ResponsesPath string          `json:"responses_path,omitempty"`
	// Modes 以默认模式开头，其余按模式名排序。
	Modes []ProviderModeInfo `json:"modes"`
}

// ProviderModeInfo 是某接入模式的默认端点；BaseURLs 缺少某协议即该模式不提供该原生端点。
type ProviderModeInfo struct {
	Mode          string            `json:"mode"`
	BaseURLs      map[string]string `json:"base_urls"`
	ProtocolRules []ProtocolRule    `json:"protocol_rules,omitempty"`
}

// BuildPlatformCatalog 按展示顺序返回全部具体平台（不含 composite）。
func BuildPlatformCatalog() PlatformCatalog {
	specs := domain.Platforms()
	catalog := PlatformCatalog{
		Platforms:           make([]PlatformCatalogEntry, 0, len(specs)),
		CompositePrecedence: domain.CompositePrecedencePlatformIDs(),
	}
	for _, spec := range specs {
		catalog.Platforms = append(catalog.Platforms, PlatformCatalogEntry{
			PlatformSpec:  spec,
			MultiProtocol: providerProfileInfo(LookupProviderProfile(spec.ID)),
		})
	}
	return catalog
}

func providerProfileInfo(profile *ProviderProfile) *ProviderProfileInfo {
	if profile == nil {
		return nil
	}
	routing := profile.Routing
	if routing == "" {
		routing = ProviderRoutingByInbound
	}
	modes := make([]string, 0, len(profile.Modes))
	for mode := range profile.Modes {
		if mode != profile.DefaultMode {
			modes = append(modes, mode)
		}
	}
	sort.Strings(modes)
	if _, ok := profile.Modes[profile.DefaultMode]; ok {
		modes = append([]string{profile.DefaultMode}, modes...)
	}
	info := &ProviderProfileInfo{
		DefaultMode:   profile.DefaultMode,
		Routing:       routing,
		ResponsesPath: profile.ResponsesPath,
		Modes:         make([]ProviderModeInfo, 0, len(modes)),
	}
	for _, mode := range modes {
		endpoints := profile.Modes[mode]
		baseURLs := make(map[string]string, len(endpoints.BaseURLs))
		for protocol, baseURL := range endpoints.BaseURLs {
			baseURLs[protocol] = baseURL
		}
		info.Modes = append(info.Modes, ProviderModeInfo{
			Mode:          mode,
			BaseURLs:      baseURLs,
			ProtocolRules: append([]ProtocolRule(nil), endpoints.ProtocolRules...),
		})
	}
	return info
}

File: frontend/src/constants/platformCatalog.ts
import { shallowRef } from 'vue'
import builtinCatalog from './platformCatalog.builtin.json'

/**
 * 平台清单：后端的平台清单（domain/platforms.go）加上多协议 API Key 供应商 profile。
 *
 * 平台选项、展示名、多协议账号表单的默认端点与分流规则都从这里读取。清单文件
 * platformCatalog.builtin.json 由后端测试 TestFrontendBuiltinPlatformCatalogInSync
 * 生成并校验与后端一致；前端随后端一同构建发布，无需运行时拉取。
 */

export type PlatformGateway = 'anthropic' | 'openai' | 'gemini' | 'antigravity'
export type ProviderRouting = 'by_inbound' | 'by_model'
export type ProviderNativeProtocol = 'chat_completions' | 'anthropic' | 'responses'

export interface ProviderProtocolRule {
  pattern: string
  protocol: string
}

export interface ProviderModeInfo {
  mode: string
  /** 以上游协议为键的官方默认基址；缺少某协议即该模式不提供该原生端点。 */
  base_urls: Partial<Record<string, string>>
  protocol_rules?: ProviderProtocolRule[]
}

export interface ProviderProfileInfo {
  default_mode: string
  routing: ProviderRouting
  responses_path?: string
  /** 默认模式在前。 */
  modes: ProviderModeInfo[]
}

export interface PlatformSpec {
  id: string
  display_name: string
  gateway: PlatformGateway
  cn_provider: boolean
  litellm_provider?: string
  multi_protocol?: ProviderProfileInfo
}

export interface PlatformCatalog {
  platforms: PlatformSpec[]
  /** 组合分组逐平台回退的顺序（先命中者生效）。 */
  composite_precedence: string[]
}

export const BUILTIN_PLATFORM_CATALOG: PlatformCatalog = builtinCatalog as PlatformCatalog

const catalogState = shallowRef<PlatformCatalog>(BUILTIN_PLATFORM_CATALOG)

/** 替换当前清单（测试用于模拟新登记的平台）。 */
export function setPlatformCatalog(catalog: PlatformCatalog): void {
  catalogState.value = catalog
}

/** 恢复内置清单（测试用）。 */
export function resetPlatformCatalog(): void {
  catalogState.value = BUILTIN_PLATFORM_CATALOG
}

/** 全部具体平台，按展示顺序。 */
export function listPlatforms(): PlatformSpec[] {
  return catalogState.value.platforms
}

export function listPlatformIds(): string[] {
  return listPlatforms().map(spec => spec.id)
}

export function getPlatformSpec(platform: string | null | undefined): PlatformSpec | undefined {
  if (!platform) return undefined
  return catalogState.value.platforms.find(spec => spec.id === platform)
}

export function isKnownPlatform(platform: string | null | undefined): boolean {
  return !!getPlatformSpec(platform)
}

/** 平台展示名；未登记的值原样返回。 */
export function platformDisplayName(platform: string | null | undefined): string {
  return getPlatformSpec(platform)?.display_name ?? platform ?? ''
}

export function compositePrecedencePlatformIds(): string[] {
  return catalogState.value.composite_precedence
}

export function platformUsesGateway(platform: string | null | undefined, gateway: PlatformGateway): boolean {
  return getPlatformSpec(platform)?.gateway === gateway
}

export function getProviderProfile(platform: string | null | undefined): ProviderProfileInfo | undefined {
  return getPlatformSpec(platform)?.multi_protocol ?? undefined
}

/** 指定接入模式的端点；未知模式回落默认模式（与后端 ProviderProfile.Endpoints 一致）。 */
export function getProviderMode(platform: string | null | undefined, mode: string | null | undefined): ProviderModeInfo | undefined {
  const profile = getProviderProfile(platform)
  if (!profile) return undefined
  const trimmed = (mode ?? '').trim()
  return (
    profile.modes.find(item => item.mode === trimmed) ??
    profile.modes.find(item => item.mode === profile.default_mode) ??
    profile.modes[0]
  )
}
